src/Http/Routing.Abstractions/src/ILiteralConstraint.cs

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Routing
+{
+    /// <summary>
+    /// Defines the contract that a class must implement in order to check if a literal value is valid for a given constraint.
+    /// <remarks>
+    /// When a parameter implements this interface, the router is able to optimize away some paths from the route table that don't match this constraint.
+    /// </remarks>
+    /// </summary>
+    public interface ILiteralConstraint : IParameterPolicy
+    {
+        /// <summary>
+        /// Determines whether the given <paramref name="literal"/> can match the constraint.
+        /// </summary>
+        /// <param name="parameterName">The parameter name we are currently evaluating.</param>
+        /// <param name="literal">The literal to test the constraint against.</param>
+        /// <returns><c>true</c> if the literal contains a valid value; otherwise, <c>false</c>.</returns>
+        bool MatchLiteral(string parameterName, string literal);
+    }
```

Comment:
Need to come up with a better name for this, in essence we want to implement this on our constraints and let people implement it too. Other names are `IParameterNodeLiteralPolicy` or `IParameterNodeBuilderPolicy`.

The idea here is that we pass literal values when we are building the DFA tree and you get to decide whether they match or not. 

We also pass you the parameter name so that you can implement policies based on the parameter name even if you don't know the values ahead of time. For example, if the parameter name is controller or page, you can choose to never add any literal from the tree to mean something like, this parameter will never match any value from the list of literals in the tree.

Reply:
I'm going to move this from the abstractions assembly to the implementation assembly, since this is mostly a DFA concern and rename it to IRouteParameterNodeFilterPolicy